README.md
# TTS ROS2

## Overview
The `tts` ROS node enables a robot to speak with a human voice by providing a Text-To-Speech service.
Out of the box this package listens to a speech topic, submits text to the Amazon Polly cloud service to generate an audio stream file,
retrieves the audio stream from Amazon Polly, and plays the audio stream via the default output device.
The nodes can be configured to use different voices as well as custom lexicons and SSML tags which enable you to control aspects of speech,
such as pronunciation, volume, pitch, speed rate, etc. A [sample ROS application] with this node,
and more details on speech customization are available within the [Amazon Polly documentation].

**Amazon Polly Summary**: Amazon Polly is a service that turns text into lifelike speech, allowing you to create applications that talk,
and build entirely new categories of speech-enabled products. Amazon Polly is a Text-to-Speech service that uses advanced deep learning technologies to synthesize speech that sounds like a human voice.
With dozens of lifelike voices across a variety of languages, you can select the ideal voice and build speech-enabled applications that work in many different countries.

### License
The source code is released under an [Apache 2.0].

**Author**: AWS RoboMaker<br/>
**Affiliation**: [Amazon Web Services (AWS)]<br/>
**Maintainer**: AWS RoboMaker, ros-contributions@amazon.com

### Supported ROS Distributions
- Dashing

### Build status
* Travis CI:
    * "master" branch [![Build Status](https://travis-ci.org/aws-robotics/tts-ros2.svg?branch=master)](https://travis-ci.org/aws-robotics/tts-ros2/branches)
    * "release-latest" branch [![Build Status](https://travis-ci.org/aws-robotics/tts-ros2.svg?branch=release-latest)](https://travis-ci.org/aws-robotics/tts-ros2/branches)
* ROS build farm:
   * ROS Dashing @ u18.04 Bionic [![Build Status](http://build.ros2.org/job/Dbin_uB64__tts__ubuntu_bionic_amd64__binary/badge/icon)](http://build.ros2.org/job/Dbin_uB64__tts__ubuntu_bionic_amd64__binary)

## Installation

### AWS Credentials
You will need to create an AWS Account and configure the credentials to be able to communicate with AWS services. You may find [AWS Configuration and Credential Files] helpful.

This node will require the following AWS account IAM role permissions:
- `polly:SynthesizeSpeech`

### Binaries
On Ubuntu you can install the latest version of this package using the following command

        sudo apt-get update
        sudo apt-get install -y ros-$ROS_DISTRO-tts

### Building from Source

To build from source you'll need to create a new workspace, clone and checkout the latest release branch of this repository, install all the dependencies, and compile. If you need the latest development features you can clone from the `master` branch instead of the latest release branch. While we guarantee the release branches are stable, __the `master` should be considered to have an unstable build__ due to ongoing development. 

- Create a ROS workspace and a source directory

        mkdir -p ~/ros-workspace/src

- Clone the package into the source directory . 

        cd ~/ros-workspace/src
        git clone https://github.com/aws-robotics/tts-ros2.git -b release-latest

- Install dependencies

        cd ~/ros-workspace 
        sudo apt-get update && rosdep update
        rosdep install --from-paths src --ignore-src -r -y
        
_Note: If building the master branch instead of a release branch you may need to also checkout and build the master branches of the packages this package depends on._

- Build the packages

        cd ~/ros-workspace && colcon build

- Configure ROS library Path

        source ~/ros-workspace/install/setup.bash
        
- Build and run the unit tests

        colcon test --packages-select tts && colcon test-result --all

#### Test on Containers/Virtual Machines

Even if your container or virtual machine does not have audio device, you can still test TTS by leveraging an audio server.

The following is an example setup on a MacBook with PulseAudio as the audio server.
If you are new to PulseAudio, you may want to read the [PulseAudio Documentation].

**Step 1: Start PulseAudio on your laptop**

After installation, start the audio server with *module-native-protocol-tcp* loaded:

    pulseaudio --load=module-native-protocol-tcp --exit-idle-time=-1 --log-target=stderr -v

Note the extra arguments `-v` and `--log-target` are used for easier troubleshooting.

**Step 2: Run TTS nodes in container**

In your container, make sure you set the right environment variables.
For example, you can start the container using `docker run -it -e PULSE_SERVER=docker.for.mac.localhost ubuntu:16.04`.

Then you will be able to run ROS nodes in the container and hear the audio from your laptop speakers.

**Troubleshooting**

If your laptop has multiple audio output devices, make sure the right one has the right volume.
This command will give you a list of output devices and tell you which one has been selected:

    pacmd list-sinks | grep -E '(index:|name:|product.name)'

## Usage

### Run the node
- **Plain text**
  - `ros2 launch tts tts.launch.py`
  - `ros2 run tts voicer 'Hello World'`

- **SSML**
  - `ros2 launch tts tts.launch.py`
  - `ros2 run tts voicer '<speak>Mary has a <amazon:effect name="whispered">little lamb.</amazon:effect></speak>' '{"text_type":"ssml"}'`


## Configuration File and Parameters
| Parameter Name | Type | Description |
| -------------- | ---- | ----------- |
| polly_action | *string* | Currently only one action named `SynthesizeSpeech` is supported. |
| text | *string* | The text to be synthesized. It can be plain text or SSML. See also `text_type`. |
| text_type | *string* | A user can choose from `text` and `ssml`. Default: `text`. |
| voice_id | *string* | The list of supported voices can be found on [official Amazon Polly document]. Default: Joanna |
| output_format | *string* | Valid formats are `ogg_vorbis`, `mp3` and `pcm`. Default: `ogg_vorbis` |
| output_path | *string* | The audio data will be saved as a local file for playback and reuse/inspection purposes. This parameter is to provide a preferred path to save the file. Default: `.` |
| sample_rate | *string* | Note `16000` is a valid sample rate for all supported formats. Default: `16000`. |


## Nodes

### polly
Polly node is the engine for the synthesizing job. It provides user-friendly yet powerful APIs so a user doesn't have to deal with technical details of AWS service calls.

#### Services
- **`polly (tts/Polly)`**

  Call the service to use Amazon Polly to synthesize the audio.

#### Reserved for future usage
- `language_code (string, default: None)`
  
  A user doesn't have to provide a language code and this is reserved for future usage.

- `lexicon_content (string, default: None)`

- `lexicon_name (string, default: None)`

- `lexicon_names (string[], default: empty)`

- `speech_mark_types (string[], default: empty)`

- `max_results (uint32, default: None)`

- `next_token (string, default: None)`

- `sns_topic_arn (string, default: None)`

- `task_id (string, default: None)`

- `task_status (string, default: iNone)`

- `output_s3_bucket_name (string, default: None)`

- `output_s3_key_prefix (string, default: None)`

- `include_additional_language_codes (bool, default: None)`

### synthesizer node

#### Services
- **`synthesizer (tts/Synthesizer)`**

  Call the service to synthesize.

#### Parameters

- **`text (string)`**

    The text to be synthesized.

- **`metadata (string, JSON format)`**

    Optional, for user to have control over how synthesis happens.

### tts node

#### Action

- **`speech`**

#### Parameters

- **`text (string)`**

  The text to be synthesized.

- **`metadata (string, JSON format)`**

  Optional, for user to have control over how synthesis happens.


## Bugs & Feature Requests
Please contact the team directly if you would like to request a feature.

Please report bugs in [Issue Tracker].


[AWS Configuration and Credential Files]: https://docs.aws.amazon.com/cli/latest/userguide/cli-config-files.html
[Amazon Polly documentation]: https://docs.aws.amazon.com/polly/latest/dg/what-is.html
[Amazon Web Services (AWS)]: https://aws.amazon.com/
[Apache 2.0]: https://aws.amazon.com/apache-2-0/
[Issue Tracker]: https://github.com/aws-robotics/tts-ros2/issues
[PulseAudio Documentation]: https://www.freedesktop.org/wiki/Software/PulseAudio/Documentation/
[official Amazon Polly document]: https://docs.aws.amazon.com/polly/latest/dg/voicelist.html
[sample ROS application]: https://github.com/aws-robotics/aws-robomaker-sample-application-voiceinteraction


tts/CHANGELOG.rst
^^^^^^^^^^^^^^^^^^^^^^^^^
Changelog for package tts
^^^^^^^^^^^^^^^^^^^^^^^^^

2.0.2 (2020-01-22)
------------------
* bump version of tts to 2.0.2 in setup.py (`#16 <https://github.com/aws-robotics/tts-ros2/issues/16>`_)
* bump package version for tts and tts_interfaces to 2.0.2 (`#15 <https://github.com/aws-robotics/tts-ros2/issues/15>`_)
* missed a change with moving voicer.py to scripts directory in https://github.com/aws-robotics/tts-ros2/pull/13
* move voicer to subdirectory similar to tts-ros1
* remove unnecessary lines of code from unit tests
* Contributors: M M, Ryan Newell

2.0.1 (2019-09-20)
------------------
* Fix voicer and bump version to 2.0.1 (`#9 <https://github.com/aws-robotics/tts-ros2/issues/9>`_)
  * Fix voicer and bump version to 2.0.1
  * Bump version of tts_interfaces to 2.0.1
* Increase wait_for_service timeout in tts_integration.py
* Contributors: AAlon

2.0.0 (2019-09-06)
------------------
* Delete CHANGELOG.rst
* Bump version to 2.0.0 and add changelog
* Add launch dependencies
* Rename python-boto3 to python3-boto3 in package.xml
* Fix imports in tests
* Fix integration test and add graceful shutdown of the server nodes.
* Merge pull request `#1 <https://github.com/aws-robotics/tts-ros2/issues/1>`_ from aws-robotics/dev
  initial commit for tts-ros2
* initial commit for tts-ros2
* Contributors: AAlon, Avishay Alon, Yuan "Forrest" Yu, y²


tts/test/test_unit_polly.py
#!/usr/bin/env python3

# Copyright (c) 2018, Amazon.com, Inc. or its affiliates. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License").
# You may not use this file except in compliance with the License.
# A copy of the License is located at
#
#  http://aws.amazon.com/apache2.0
#
# or in the "license" file accompanying this file. This file is distributed
# on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
# express or implied. See the License for the specific language governing
# permissions and limitations under the License.

from unittest.mock import patch, MagicMock
import unittest


class TestPolly(unittest.TestCase):

    def setUp(self):
        self.patchers = {
            s: patch(s) for s in
            (
                'rclpy.init',
                'rclpy.spin',
                'rclpy.shutdown',
            )
        }
        self.mocks = {s: p.start() for s, p in self.patchers.items()}

    def tearDown(self):
        for p in self.patchers.values():
            p.stop()

    @patch('tts.services.amazonpolly.Session')
    def test_init(self, boto3_session_class_mock):
        from tts.services.amazonpolly import AmazonPolly
        AmazonPolly()

        self.assertGreater(boto3_session_class_mock.call_count, 0)
        boto3_session_class_mock.return_value.client.assert_called_with('polly')

    @patch('tts.services.amazonpolly.Session')
    def test_defaults(self, boto3_session_class_mock):
        from tts.services.amazonpolly import AmazonPolly
        polly = AmazonPolly()

        self.assertGreater(boto3_session_class_mock.call_count, 0)
        boto3_session_class_mock.return_value.client.assert_called_with('polly')

        self.assertEqual('text', polly.default_text_type)
        self.assertEqual('pcm', polly.default_output_format)
        self.assertEqual('Joanna', polly.default_voice_id)
        self.assertEqual('.', polly.default_output_folder)
        self.assertEqual('output', polly.default_output_file_basename)

    @patch('tts.services.amazonpolly.Session')
    def test_good_synthesis_with_default_args(self, boto3_session_class_mock):
        boto3_session_obj_mock = MagicMock()
        boto3_polly_obj_mock = MagicMock()
        boto3_polly_response_mock = MagicMock()
        audio_stream_mock = MagicMock()
        fake_audio_stream_data = 'I am audio.'
        fake_audio_content_type = 'super tts'
        fake_boto3_polly_response_metadata = {'foo': 'bar'}

        boto3_session_class_mock.return_value = boto3_session_obj_mock
        boto3_session_obj_mock.client.return_value = boto3_polly_obj_mock
        boto3_polly_obj_mock.synthesize_speech.return_value = boto3_polly_response_mock
        audio_stream_mock.read.return_value = fake_audio_stream_data.encode('utf-8')
        d = {
            'AudioStream': audio_stream_mock,
            'ContentType': fake_audio_content_type,
            'ResponseMetadata': fake_boto3_polly_response_metadata
        }
        boto3_polly_response_mock.__contains__.side_effect = d.__contains__
        boto3_polly_response_mock.__getitem__.side_effect = d.__getitem__

        from tts.services.amazonpolly import AmazonPolly
        polly_under_test = AmazonPolly()

        self.assertGreater(boto3_session_class_mock.call_count, 0)
        boto3_session_obj_mock.client.assert_called_with('polly')

        res = polly_under_test.synthesize(output_format='ogg_vorbis', text='hello')

        expected_synthesize_speech_kwargs = {
            'LexiconNames': [],
            'OutputFormat': 'ogg_vorbis',
            'SampleRate': '22050',
            'SpeechMarkTypes': [],
            'Text': 'hello',
            'TextType': 'text',
            'VoiceId': 'Joanna',
        }
        boto3_polly_obj_mock.synthesize_speech.assert_called_with(**expected_synthesize_speech_kwargs)

        from tts_interfaces.srv import Polly
        self.assertTrue(isinstance(res, Polly.Response))

        import json
        j = json.loads(res.result)
        observed_audio_file_content = open(j['Audio File']).read()
        self.assertEqual(fake_audio_stream_data, observed_audio_file_content)

        self.assertEqual(fake_audio_content_type, j['Audio Type'])
        self.assertEqual(str(fake_boto3_polly_response_metadata), j['Amazon Polly Response Metadata'])

    @patch('tts.services.amazonpolly.Session')
    def test_polly_raises(self, boto3_session_class_mock):
        boto3_session_obj_mock = MagicMock()
        boto3_polly_obj_mock = MagicMock()
        boto3_polly_response_mock = MagicMock()
        audio_stream_mock = MagicMock()
        fake_audio_stream_data = 'I am audio.'
        fake_audio_content_type = 'super voice'
        fake_boto3_polly_response_metadata = {'foo': 'bar'}

        boto3_session_class_mock.return_value = boto3_session_obj_mock
        boto3_session_obj_mock.client.return_value = boto3_polly_obj_mock
        boto3_polly_obj_mock.synthesize_speech.side_effect = RuntimeError('Amazon Polly Exception')
        audio_stream_mock.read.return_value = fake_audio_stream_data
        d = {
            'AudioStream': audio_stream_mock,
            'ContentType': fake_audio_content_type,
            'ResponseMetadata': fake_boto3_polly_response_metadata
        }
        boto3_polly_response_mock.__contains__.side_effect = d.__contains__
        boto3_polly_response_mock.__getitem__.side_effect = d.__getitem__

        from tts.services.amazonpolly import AmazonPolly
        polly_under_test = AmazonPolly()

        self.assertGreater(boto3_session_class_mock.call_count, 0)
        boto3_session_obj_mock.client.assert_called_with('polly')

        res = polly_under_test.synthesize(text='hello')

        expected_synthesize_speech_kwargs = {
            'LexiconNames': [],
            'OutputFormat': 'pcm',
            'SampleRate': '16000',
            'SpeechMarkTypes': [],
            'Text': 'hello',
            'TextType': 'text',
            'VoiceId': 'Joanna',
        }
        boto3_polly_obj_mock.synthesize_speech.assert_called_with(**expected_synthesize_speech_kwargs)

        from tts_interfaces.srv import Polly
        self.assertTrue(isinstance(res, Polly.Response))

        import json
        j = json.loads(res.result)
        self.assertTrue('Exception' in j)
        self.assertTrue('Traceback' in j)

    @patch('tts.services.amazonpolly.AmazonPollyNode')
    def test_cli(self, amazon_polly_class_mock):
        import sys
        with patch.object(sys, 'argv', ['amazonpolly.py', '-n', 'pollynode']):
            from tts.services import amazonpolly
            amazonpolly.main()
            self.assertGreater(amazon_polly_class_mock.call_count, 0)


if __name__ == '__main__':
    unittest.main()


tts/tts/voicer.py
# Copyright 2016 Open Source Robotics Foundation, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from tts_interfaces.srv import Synthesizer

import rclpy
from rclpy.node import Node
import sys
import json
import os
import time

import gi
gi.require_version('Gst', '1.0')
from gi.repository import Gst


class Voicer(Node):
    def __init__(self):
        super().__init__('tts_node')
        self.synthesizer = self.create_client(Synthesizer, 'synthesizer')
        self.logger = self.get_logger()
        Gst.init(None)

    def handle_result(self, synthesizer_response):
        result = synthesizer_response.result
        try:
            r = json.loads(result)
        except Exception as e:
            s = 'Expecting JSON from synthesizer but got {}'.format(result)
            self.logger.error('{}. Exception: {}'.format(s, e))
            return

        result = ''

        if 'Audio File' in r:
            audio_file = r['Audio File']
            self.logger.info('Will play {}'.format(audio_file))
            self.play(audio_file)
            result = audio_file

        if 'Exception' in r:
            result = '[ERROR] {}'.format(r)
            self.logger.error(result)

        return result

    def play(self, filename):
        self.logger.info('using gstreamer to play the audio')

        playbin = Gst.ElementFactory.make('playbin', 'player')

        bus = playbin.get_bus()

        playbin.props.uri = 'file://' + os.path.abspath(filename)
        time.sleep(0.5)  # sometimes gst needs time to get ready for unknown reasons
        set_result = playbin.set_state(Gst.State.PLAYING)
        if set_result != Gst.StateChangeReturn.ASYNC:
            raise RuntimeError("gstreamer error: playbin.set_state returned " + repr(set_result))

        bus.poll(Gst.MessageType.EOS, Gst.CLOCK_TIME_NONE)
        playbin.set_state(Gst.State.NULL)

    def speak(self):
        self.req = Synthesizer.Request()
        self.req.text = sys.argv[1] if len(sys.argv) > 1 else 'I got no idea.'
        self.req.metadata = sys.argv[2] if len(sys.argv) > 2 else ''

        while not self.synthesizer.wait_for_service(timeout_sec=1.0):
            self.logger.warn('service not available, waiting again...')

        future = self.synthesizer.call_async(self.req)
        future.add_done_callback(self.on_synthesize_done)
        return future

    def on_synthesize_done(self, future):
        if future.result():
            self.logger.info('Result: %s' % str(future.result()))
            self.handle_result(future.result())
            self.logger.info('Done speaking {}'.format(self.req.text))
        else:
            self.logger.error('Exception while calling service: %r' % future.exception())


def main():
    rclpy.init()
    node = Voicer()
    future = node.speak()
    # Spin until future completes, then once more to execute its completion callback
    rclpy.spin_until_future_complete(node, future)
    rclpy.spin_once(node)
    node.destroy_node()
    rclpy.shutdown()


if __name__ == '__main__':
    main()


tts/tts/services/amazonpolly.py
# Copyright (c) 2018, Amazon.com, Inc. or its affiliates. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License").
# You may not use this file except in compliance with the License.
# A copy of the License is located at
#
#  http://aws.amazon.com/apache2.0
#
# or in the "license" file accompanying this file. This file is distributed
# on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
# express or implied. See the License for the specific language governing
# permissions and limitations under the License.

import json
import os
import sys
import wave
import traceback
from boto3 import Session
from botocore.session import get_session
from botocore.exceptions import UnknownServiceError
from contextlib import closing
from optparse import OptionParser

from tts_interfaces.srv import Polly

import rclpy
from rclpy.node import Node


class AmazonPolly:
    """
    A TTS engine that can be used in two different ways.

    Usage
    -----

    1. It can run as a ROS service node.

    Start a polly node::

        $ ros2 run tts polly_server

    Call the service from command line::

        $ ros2 service call /polly tts_interfaces/Polly "{polly_action: 'SynthesizeSpeech', text: 'hello'}"

    Call the service programmatically::

        from tts_interfaces.srv import Polly

        rclpy.init()
        node = rclpy.create_node('ClientNode')
        client = node.create_client(Polly, 'polly')
        client.wait_for_service()
        text = 'Mary has a little lamb, little lamb, little lamb.'
        request = Polly.Request(polly_action='SynthesizeSpeech', text=text)
        future = client.call_async(request)

        while rclpy.ok():
            rclpy.spin_once(node)
            if future.done():
                res = future.result()
                ...
        ...
        node.destroy_node()
        rclpy.shutdown()

    2. It can also be used as a normal python class::

        AmazonPolly().synthesize(text='hi polly')

    Polly.Request supports many parameters, but the majority of the users can safely ignore most of them and just
    use the vanilla version which involves only one argument, ``text``.

    If in some use cases more control is needed, SSML will come handy. Example::

        AmazonPolly().synthesize(
            text='<speak>Mary has a <amazon:effect name="whispered">little lamb.</amazon:effect></speak>',
            text_type='ssml'
        )

    A user can also control the voice, output format and so on. Example::

        AmazonPolly().synthesize(
            text='<speak>Mary has a <amazon:effect name="whispered">little lamb.</amazon:effect></speak>',
            text_type='ssml',
            voice_id='Joey',
            output_format='mp3',
            output_path='/tmp/blah'
        )


    Service Parameters
    ------------------

    Among the parameters defined in Polly.srv, the following are supported while others are reserved for future.

    * polly_action : currently only ``SynthesizeSpeech`` is supported
    * text : the text to speak
    * text_type : can be either ``text`` (default) or ``ssml``
    * voice_id : any voice id supported by Amazon Polly, default is Joanna
    * output_format : pcm (default), mp3 or ogg
    * output_path : where the audio file is saved
    * sample_rate : default is 16000 for pcm or 22050 for mp3 and ogg

    The following are the reserved ones. Note that ``language_code`` is rarely needed (this may seem counter-intuitive).
    See official Amazon Polly documentation for details (link can be found below).

    * language_code
    * lexicon_content
    * lexicon_name
    * lexicon_names
    * speech_mark_types
    * max_results
    * next_token
    * sns_topic_arn
    * task_id
    * task_status
    * output_s3_bucket_name
    * output_s3_key_prefix
    * include_additional_language_codes


    Links
    -----

    Amazon Polly documentation: https://docs.aws.amazon.com/polly/latest/dg/API_SynthesizeSpeech.html

    """

    def __init__(self, aws_access_key_id=None, aws_secret_access_key=None, aws_session_token=None, region_name=None,
                 logger=None):
        if region_name is None:
            region_name = 'us-west-2'

        # for maximum flexibility, logger can be None, in which case nothing will be logged
        self.logger = logger

        self.polly = self._get_polly_client(aws_access_key_id, aws_secret_access_key, aws_session_token, region_name)
        self.default_text_type = 'text'
        self.default_voice_id = 'Joanna'
        self.default_output_format = 'pcm'
        self.default_output_folder = '.'
        self.default_output_file_basename = 'output'

    def loginfo(self, s):
        if self.logger:
            self.logger.info(s)

    def logerr(self, s):
        if self.logger:
            self.logger.error(s)

    def _get_polly_client(self, aws_access_key_id=None, aws_secret_access_key=None, aws_session_token=None,
                          region_name=None, with_service_model_patch=False):
        # Note we get a new botocore session each time this function is called
        # to avoid potential problems caused by inner state of the session.
        botocore_session = get_session()

        if with_service_model_patch:
            # Older versions of botocore don't have polly. We can possibly fix it by appending
            # extra path with polly service model files to the search path.
            current_dir = os.path.dirname(os.path.abspath(__file__))
            service_model_path = os.path.join(current_dir, 'data', 'models')
            botocore_session.set_config_variable('data_path', service_model_path)
            self.loginfo('patching service model data path: {}'.format(service_model_path))

        botocore_session.user_agent_extra = self._generate_user_agent_suffix()

        session = Session(aws_access_key_id=aws_access_key_id, aws_secret_access_key=aws_secret_access_key,
                          aws_session_token=aws_session_token, region_name=region_name,
                          botocore_session=botocore_session)

        try:
            return session.client("polly")
        except UnknownServiceError:
            # the first time we reach here, we try to fix the problem
            if not with_service_model_patch:
                return self._get_polly_client(aws_access_key_id, aws_secret_access_key, aws_session_token, region_name,
                                              with_service_model_patch=True)
            else:
                # we have tried our best, time to panic
                self.logerr('Amazon Polly is not available. Please install the latest boto3.')
                raise

    def _generate_user_agent_suffix(self):
        exec_env = 'AWS_RoboMaker'
        ros_distro = 'ROS2'
        ros_version = 'Unknown_ROS_VERSION'
        return 'exec-env/{} ros-{}/{}'.format(exec_env, ros_distro, ros_version)

    def _pcm2wav(self, audio_data, wav_filename, sample_rate):
        # per Amazon Polly official doc, the pcm in a signed 16-bit, 1 channel (mono), little-endian format.
        wavf = wave.open(wav_filename, 'w')
        wavf.setframerate(int(sample_rate))
        wavf.setnchannels(1)  # 1 channel
        wavf.setsampwidth(2)  # 2 bytes == 16 bits
        wavf.writeframes(audio_data)
        wavf.close()

    def _make_audio_file_fullpath(self, output_path, output_format):
        """
        Make a full path for audio file based on given output path and format.

        If ``output_path`` doesn't have a path, current path is used.

        :param output_path: the output path received
        :param output_format: the audio format, e.g., mp3, ogg_vorbis, pcm
        :return: a full path for the output audio file. File ext will be constructed from audio format.
        """
        head, tail = os.path.split(output_path)
        if not head:
            head = self.default_output_folder
        if not tail:
            tail = self.default_output_file_basename

        file_ext = {'pcm': '.wav', 'mp3': '.mp3', 'ogg_vorbis': '.ogg'}[output_format.lower()]
        if not tail.endswith(file_ext):
            tail += file_ext

        return os.path.realpath(os.path.join(head, tail))

    def _synthesize_speech_and_save(self, request):
        """
        Call Amazon Polly and writes the returned audio data to a local file.

        To make it practical, three things will be returned in a JSON form string, which are audio file path,
        audio type and Amazon Polly response metadata.

        If the Amazon Polly call fails, audio file name will be an empty string and audio type will be "N/A".

        Please see https://boto3.readthedocs.io/reference/services/polly.html#Polly.Client.synthesize_speech
        for more details on Amazon Polly API.

        :param request: an instance of Polly.Request
        :return: a string in JSON form with two attributes, "Audio File" and "Amazon Polly Response".
        """
        kws = {
            'LexiconNames': request.lexicon_names if request.lexicon_names else [],
            'OutputFormat': request.output_format if request.output_format else self.default_output_format,
            'SampleRate': request.sample_rate,
            'SpeechMarkTypes': request.speech_mark_types if request.speech_mark_types else [],
            'Text': request.text,
            'TextType': request.text_type if request.text_type else self.default_text_type,
            'VoiceId': request.voice_id if request.voice_id else self.default_voice_id
        }

        if not kws['SampleRate']:
            kws['SampleRate'] = '16000' if kws['OutputFormat'].lower() == 'pcm' else '22050'

        self.loginfo('Amazon Polly Request: {}'.format(kws))
        response = self.polly.synthesize_speech(**kws)
        self.loginfo('Amazon Polly Response: {}'.format(response))

        if "AudioStream" in response:
            audiofile = self._make_audio_file_fullpath(request.output_path, kws['OutputFormat'])
            self.loginfo('will save audio as {}'.format(audiofile))

            with closing(response["AudioStream"]) as stream:
                if kws['OutputFormat'].lower() == 'pcm':
                    self._pcm2wav(stream.read(), audiofile, kws['SampleRate'])
                else:
                    with open(audiofile, "wb") as f:
                        f.write(stream.read())

            audiotype = response['ContentType']
        else:
            audiofile = ''
            audiotype = 'N/A'

        return json.dumps({
            'Audio File': audiofile,
            'Audio Type': audiotype,
            'Amazon Polly Response Metadata': str(response['ResponseMetadata'])
        })

    def _dispatch(self, request):
        """
        Amazon Polly supports a number of APIs. This will call the right one based on the content of request.

        Currently "SynthesizeSpeech" is the only recognized action. Basically this method just delegates the work
        to ``self._synthesize_speech_and_save`` and returns the result as is. It will simply raise if a different
        action is passed in.

        :param request: an instance of Polly.Request
        :return: whatever returned by the delegate
        """
        actions = {
            'SynthesizeSpeech': self._synthesize_speech_and_save
            # ... more actions could go in here ...
        }

        if request.polly_action not in actions:
            raise RuntimeError('bad or unsupported Amazon Polly action: "' + request.polly_action + '".')

        return actions[request.polly_action](request)

    def node_request_handler(self, request, response):
        """
        Process service request. The callback function.

        It never raises. If anything unexpected happens, it will return a Polly.Response with details of the exception.

        :param request: an instance of Polly.Request
        :param response: an instance of Polly.Response
        :return: a Polly.Response
        """
        self.loginfo('Amazon Polly Request: {}'.format(request))

        try:
            result = self._dispatch(request)
            self.loginfo('will return {}'.format(result))
            response.result = result
            return response
        except Exception as e:
            current_dir = os.path.dirname(os.path.abspath(__file__))  # todo: make use of pkg_resources
            exc_type = sys.exc_info()[0]

            # not using `issubclass(exc_type, ConnectionError)` for the condition below because some versions
            # of urllib3 raises exception when doing `from requests.exceptions import ConnectionError`
            error_ogg_filename = 'connerror.ogg' if 'ConnectionError' in exc_type.__name__ else 'error.ogg'

            error_details = {
                'Audio File': os.path.join(current_dir, 'data', error_ogg_filename),
                'Audio Type': 'ogg',
                'Exception': {
                    'Type': str(exc_type),
                    'Module': exc_type.__module__,
                    'Name': exc_type.__name__,
                    'Value': str(e),
                },
                'Traceback': traceback.format_exc()
            }

            error_str = json.dumps(error_details)
            self.logerr(error_str)
            response.result = error_str
            return response

    def synthesize(self, **kws):
        """
        Call this method if you want to use polly but don't want to start a node.

        :param kws: input as defined in Polly.srv
        :return: a string in JSON form with detailed information, success or failure
        """
        request = Polly.Request(polly_action='SynthesizeSpeech', **kws)
        response = Polly.Response()
        return self.node_request_handler(request, response)


class AmazonPollyNode(Node):

    def __init__(self, node_name, service_name):
        super().__init__(node_name)
        self.polly = AmazonPolly(logger=self.get_logger())
        self.srv = self.create_service(Polly, service_name, self.polly_callback)

    def polly_callback(self, request, response):
        return self.polly.node_request_handler(request, response)


def main():
    usage = '''usage: %prog [options]
    '''

    parser = OptionParser(usage)

    parser.add_option("-n", "--node-name", dest="node_name", default='polly_node',
                      help="name of the ROS node",
                      metavar="NODE_NAME")
    parser.add_option("-s", "--service-name", dest="service_name", default='polly',
                      help="name of the ROS service",
                      metavar="SERVICE_NAME")

    (options, args) = parser.parse_args()

    node_name = options.node_name
    service_name = options.service_name

    rclpy.init()

    node = AmazonPollyNode(node_name, service_name)

    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        print('Shutting down polly server')
    except BaseException:
        print('Exception in polly server:', file=sys.stderr)
        raise
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main()
